File: code/bushbuck.R
set.seed(20240521)
library(jpeg)
library(tidyverse)
library(dplyr)
library(Momocs)
library(fdasrvf)

path <- paste0("./data/bushbuckcomplete/")
file_list_BW_bushbuck <- list.files(path, recursive = TRUE, full.names = TRUE)

start <- Sys.time()
import_BW <- function(x){import_jpg(x)[[1]]}
teeth_BW_bushbuck <- lapply(as.list(file_list_BW_bushbuck), import_BW)
names(teeth_BW_bushbuck) <- unlist(lapply(strsplit(file_list_BW_bushbuck,"/"), function(x){x[[length(x)]]}))
names(teeth_BW_bushbuck) <- substring(names(teeth_BW_bushbuck),1,nchar(names(teeth_BW_bushbuck))-4)
end <- Sys.time()
end - start

save(teeth_BW_bushbuck, file = "./data/teeth_BW_bushbuck.RData")

#Export data to matlab

#Now do data prep for matlab
#Makes all the teeth have the same number of points
make_same_num_points <- function(x, N = 500){
  out <- resamplecurve(t(x),N)
  return(out)
}

load("./data/teeth_BW_bushbuck.RData")
#Make all teeth have 500 points
teeth_BW_bushbuck <- lapply(teeth_BW_bushbuck, make_same_num_points)

write.csv(do.call(rbind,teeth_BW_bushbuck),file = paste0("./data/matlab/data_bushbuck.csv"), row.names = FALSE)
write.csv(do.call(rbind,teeth_BW_bushbuck[c("CD5399fixed.","CD5410fixed.")]),file = paste0("./data/matlab/data_bushbuck_LM2.csv"), row.names = FALSE)
write.csv(do.call(rbind,teeth_BW_bushbuck[c("UW 88 519a","SK 4261.", "CD309fixedB")]),file = paste0("./data/matlab/data_bushbuck_LM3.csv"), row.names = FALSE)
write.csv(do.call(rbind,teeth_BW_bushbuck[c("CD19949.")]),file = paste0("./data/matlab/data_bushbuck_UM2.csv"), row.names = FALSE)


# %CD 309 LM3
# %CD 5399 LM2
# %CD 5410 LM2
# %CD 19949 UM2
# %519 LM3
# %SK1 LM3


#####Compute distances in matlab with bushbuck.m script

#####MDS plot for LM3
set.seed(20240521)
load("./data/teethdata_scriptus_pricei.RData")

#for (toothtype in c("LM3")){print(toothtype)
toothtype <- "LM3"
  
  n_scriptus <- length(data[[toothtype]][["scriptus"]])
  n_pricei <- length(data[[toothtype]][["pricei"]])
  
  class <- c(rep("T. scriptus",n_scriptus),rep("T. pricei",n_pricei))
  id <- c(names(data[[toothtype]][["scriptus"]]),names(data[[toothtype]][["pricei"]]),"UW 88 519a","SK 4261", "CD309B")
  
  #Run this script first in matlab: bushbuck.m
  
  #Pariwise distances for LM3
  ddd <- read.csv(paste0("./data/matlab/pairwise_distances_bushbuck_LM3.csv"), header = FALSE)
  ddd <- as.matrix(ddd)
  
  image(ddd)
  
  
  mdsdat <- data.frame(x = cmdscale(ddd)[,1], 
                       y = cmdscale(ddd)[,2],
                       species = c(class,"UW 88 519a","SK 4261.", "CD309B"), 
                       id = id)
  mdsdat <- data.frame(x = cmdscale(ddd)[,1], 
                       y = cmdscale(ddd)[,2],
                       species = c(class,rep("Fossil",3)), 
                       id = id)
library(ggrepel) 
  png(file = "./MDS_LM3_ShapeOnly.png", h = 6, w = 6, units = "in", res = 300)
  ggplot(aes(x = x, y = y, col = species), data = mdsdat) +
    geom_point(size = 1) +
    theme_bw() +
    geom_text_repel(aes(label = id),color = "#F8766D", size = 2, nudge_x = 0.02, data = mdsdat %>% filter(species == "Fossil")) + ggtitle("LM3 - Shape only") + scale_color_manual(values=c(Fossil = "#F8766D",`T. pricei` = "#619CFF",`T. scriptus` = "#00BA38")) + xlab("") + ylab("")
  dev.off()
  #library(scales)
  #show_col(hue_pal()(3))  
  
  # %CD 309 LM3
  # %CD 5399 LM2
  # %CD 5410 LM2
  # %CD 19949 UM2
  # %519 LM3
  # %SK1 LM3
  
  
  
  ###################################
  #Pariwise distances for LM2
  ###################################
  set.seed(20240521)
  load("./data/teethdata_scriptus_pricei.RData")
  
  toothtype <- "LM2"
  
  n_scriptus <- length(data[[toothtype]][["scriptus"]])
  n_pricei <- length(data[[toothtype]][["pricei"]])
  
  class <- c(rep("scriptus",n_scriptus),rep("pricei",n_pricei))
  id <- c(names(data[[toothtype]][["scriptus"]]),names(data[[toothtype]][["pricei"]]),"CD19949","CD5399","CD5410")
  
  
  ddd <- read.csv(paste0("./data/matlab/pairwise_distances_bushbuck_LM2.csv"), header = FALSE)
  ddd <- as.matrix(ddd)
  
  image(ddd)
  
  
  mdsdat <- data.frame(x = cmdscale(ddd)[,1], 
                       y = cmdscale(ddd)[,2],
                       species = c(class,"CD19949","CD5399fixed.","CD5410fixed."),
                       id = id)
  
  mdsdat <- data.frame(x = cmdscale(ddd)[,1], 
                       y = cmdscale(ddd)[,2],
                       species = c(class,rep("fossil",3)), 
                       id = id)
  mdsdat <- mdsdat %>% filter(id != "CD19949")
  
  library(ggrepel)                       
  ggplot(aes(x = x, y = y, col = species), data = mdsdat) + geom_point(size = 0.5) + theme_bw() + geom_text_repel(label = mdsdat$id, size = 2, nudge_x = 0.02) + ggtitle("LM2 - Shape only")
  
  png(file = "./MDS_LM2_ShapeOnly.png", h = 6, w = 6, units = "in", res = 300)
  ggplot(aes(x = x, y = y, col = species), data = mdsdat) +
    geom_point(size = 1) +
    theme_bw() +
    geom_text_repel(aes(label = id),color = "#F8766D", size = 2, nudge_x = 0.02, data = mdsdat %>% filter(species == "fossil")) + ggtitle("LM2 - Shape only") + scale_color_manual(values=c(fossil = "#F8766D",pricei = "#619CFF",scriptus = "#00BA38")) + xlab("") + ylab("")
  dev.off()
  
  
  ###################################
  #Pariwise distances for UM2
  ###################################
  set.seed(20240521)
  load("./data/teethdata_scriptus_pricei.RData")
  
  toothtype <- "UM2"
  
  n_scriptus <- length(data[[toothtype]][["scriptus"]])
  n_pricei <- length(data[[toothtype]][["pricei"]])
  
  class <- c(rep("scriptus",n_scriptus),rep("pricei",n_pricei))
  id <- c(names(data[[toothtype]][["scriptus"]]),names(data[[toothtype]][["pricei"]]),"CD19949","CD5399","CD5410")
  
  
  ddd <- read.csv(paste0("./data/matlab/pairwise_distances_bushbuck_UM2.csv"), header = FALSE)
  ddd <- as.matrix(ddd)
  
  image(ddd)
  
  
  mdsdat <- data.frame(x = cmdscale(ddd)[,1], 
                       y = cmdscale(ddd)[,2],
                       species = c(class,"CD19949","CD5399fixed.","CD5410fixed."),
                       id = id)
  
  mdsdat <- data.frame(x = cmdscale(ddd)[,1], 
                       y = cmdscale(ddd)[,2],
                       species = c(class,rep("fossil",3)), 
                       id = id)
  mdsdat <- mdsdat %>% filter(id != "CD19949")
  
  
  
  library(ggrepel)                       
  ggplot(aes(x = x, y = y, col = species), data = mdsdat) + geom_point(size = 0.5) + theme_bw() + geom_text_repel(label = mdsdat$id, size = 2, nudge_x = 0.02) + ggtitle("LM2 - Shape only")
  
  png(file = "./MDS_LM2_ShapeOnly.png", h = 6, w = 6, units = "in", res = 300)
  ggplot(aes(x = x, y = y, col = species), data = mdsdat) +
    geom_point(size = 1) +
    theme_bw() +
    geom_text_repel(aes(label = id),color = "#F8766D", size = 2, nudge_x = 0.02, data = mdsdat %>% filter(species == "fossil")) + ggtitle("LM2 - Shape only") + scale_color_manual(values=c(fossil = "#F8766D",pricei = "#619CFF",scriptus = "#00BA38")) + xlab("") + ylab("")
  dev.off()
  
  
  
#####Compute distances for partial shapes.  
#### i did this in the teeth-scriptus-pricei-size repo
  

File: code/matlab/bushbuck.m
wd = "/Users/gregorymatthews/Dropbox"
cd /Users/gregorymatthews/Dropbox/teeth-scriptus-pricei/code/matlab

% get data from bushbuck
data_bushbuck = readtable(wd+"/teeth-scriptus-pricei/data/matlab/data_bushbuck_LM3.csv");
        %get the number of rows and cols
        n_rows = size(data_bushbuck,1);
        n_cols = size(data_bushbuck,2);
        
        %blank array storage for matrix for each image
        teeth_data_bushbuck = zeros(2,100,n_rows/2);
        
        
        %loop to assign each tooth's coordinates to the appropriate spot in the
        %array
        for i=1:2:n_rows
            %get index for the image we are on 
            j = find((1:2:n_rows)==i);
            
            %assign matrix for image j 
             X = data_bushbuck{i:(i+1), 2:n_cols};
            
            %resample so all of the curves have 100 points
            teeth_data_bushbuck(:,:,j) = ReSampleCurve(X,100);
        end



%Get the LM3's fromscriptus and pricei
toothtype = {"LM3"};
species = {"pricei", "scriptus"};
for t=1
  data_scriptus = readtable(wd + "/teeth-scriptus-pricei/data/matlab/data_"+toothtype(t)+"_scriptus.csv")
  data_pricei = readtable(wd + "/teeth-scriptus-pricei/data/matlab/data_"+toothtype(t)+"_pricei.csv")
  data=[data_scriptus; data_pricei] 
        
        %get the number of rows and cols
        n_rows = size(data,1);
        n_cols = size(data,2);
        
        %blank array storage for matrix for each image
        teeth_data = zeros(2,100,n_rows/2);
        
        
        %loop to assign each tooth's coordinates to the appropriate spot in the
        %array
        for i=1:2:n_rows
            %get index for the image we are on 
            j = find((1:2:n_rows)==i);
            
            %assign matrix for image j 
             X = data{i:(i+1), 2:n_cols};
            
            %resample so all of the curves have 100 points
            teeth_data(:,:,j) = ReSampleCurve(X,100);
        end        

    end



load(wd + "/teeth-scriptus-pricei/data/matlab/pairwise_distances_"+toothtype(t)+".mat","ddd")

size(ddd,1)
ddd_bushbuck_LM3 = ddd;

for i = 1:size(ddd,1);
    disp(i);
    for j = 1:size(teeth_data_bushbuck,3);
        disp(j);
        ddd_bushbuck_LM3(i,size(ddd,1)+j)  = GeodesicElasticClosed(teeth_data(:,:,i),teeth_data_bushbuck(:,:,j));     
        ddd_bushbuck_LM3(size(ddd,1)+j,i) = ddd_bushbuck_LM3(i,size(ddd,1)+j);
    end
end

for i = 1:size(teeth_data_bushbuck,3);
    disp(i);
    for j = 1:size(teeth_data_bushbuck,3);
        disp(j);
        ddd_bushbuck_LM3(size(ddd,1)+i,size(ddd,1)+j)  = GeodesicElasticClosed(teeth_data_bushbuck(:,:,i),teeth_data_bushbuck(:,:,j));     
        ddd_bushbuck_LM3(size(ddd,1)+j,size(ddd,1)+i) = ddd_bushbuck_LM3(size(ddd,1)+i,size(ddd,1)+j)
        
    end
end

save(wd + "/teeth-scriptus-pricei/data/matlab/pairwise_distances_bushbuck_LM3.mat","ddd_bushbuck_LM3")
  csvwrite(wd + "/teeth-scriptus-pricei/data/matlab/pairwise_distances_bushbuck_LM3.csv",ddd_bushbuck_LM3)





  %%%%%%%%%%%%%%%%%%%
%LM2
 %%%%%%%%%%%%%%%%%%%%%
 % get data from bushbuck
data_bushbuck = readtable(wd+"/teeth-scriptus-pricei/data/matlab/data_bushbuck_LM2.csv");
        %get the number of rows and cols
        n_rows = size(data_bushbuck,1);
        n_cols = size(data_bushbuck,2);
        
        %blank array storage for matrix for each image
        teeth_data_bushbuck = zeros(2,100,n_rows/2);
        
        
        %loop to assign each tooth's coordinates to the appropriate spot in the
        %array
        for i=1:2:n_rows
            %get index for the image we are on 
            j = find((1:2:n_rows)==i);
            
            %assign matrix for image j 
             X = data_bushbuck{i:(i+1), 2:n_cols};
            
            %resample so all of the curves have 100 points
            teeth_data_bushbuck(:,:,j) = ReSampleCurve(X,100);
        end



%Get the LM2's froms criptus and pricei
toothtype = {"LM2"};
species = {"pricei", "scriptus"};
for t=1
  data_scriptus = readtable(wd + "/teeth-scriptus-pricei/data/matlab/data_"+toothtype(t)+"_scriptus.csv");
  data_pricei = readtable(wd + "/teeth-scriptus-pricei/data/matlab/data_"+toothtype(t)+"_pricei.csv");
  data=[data_scriptus; data_pricei];
        
        %get the number of rows and cols
        n_rows = size(data,1);
        n_cols = size(data,2);
        
        %blank array storage for matrix for each image
        teeth_data = zeros(2,100,n_rows/2);
        
        
        %loop to assign each tooth's coordinates to the appropriate spot in the
        %array
        for i=1:2:n_rows
            %get index for the image we are on 
            j = find((1:2:n_rows)==i);
            
            %assign matrix for image j 
             X = data{i:(i+1), 2:n_cols};
            
            %resample so all of the curves have 100 points
            teeth_data(:,:,j) = ReSampleCurve(X,100);
        end        

    end



load(wd + "/teeth-scriptus-pricei/data/matlab/pairwise_distances_"+toothtype(t)+".mat","ddd");

size(ddd,1)
ddd_bushbuck_LM2 = ddd;

for i = 1:size(ddd,1);
    for j = 1:size(teeth_data_bushbuck,3);
        ddd_bushbuck_LM2(i,size(ddd,1)+j)  = GeodesicElasticClosed(teeth_data(:,:,i),teeth_data_bushbuck(:,:,j));     
        ddd_bushbuck_LM2(size(ddd,1)+j,i) = ddd_bushbuck_LM2(i,size(ddd,1)+j);
    end
end

for i = 1:size(teeth_data_bushbuck,3);
    for j = 1:size(teeth_data_bushbuck,3);
        ddd_bushbuck_LM2(size(ddd,1)+i,size(ddd,1)+j)  = GeodesicElasticClosed(teeth_data_bushbuck(:,:,i),teeth_data_bushbuck(:,:,j));     
        ddd_bushbuck_LM2(size(ddd,1)+j,size(ddd,1)+i) = ddd_bushbuck_LM2(size(ddd,1)+i,size(ddd,1)+j)
        
    end
end

save(wd + "/teeth-scriptus-pricei/data/matlab/pairwise_distances_bushbuck_LM2.mat","ddd_bushbuck_LM2")
  csvwrite(wd + "/teeth-scriptus-pricei/data/matlab/pairwise_distances_bushbuck_LM2.csv",ddd_bushbuck_LM2)


  %%%%%%%%%%%%%%%%%%%
%UM2
 %%%%%%%%%%%%%%%%%%%%%
 % get data from bushbuck
data_bushbuck = readtable(wd+"/teeth-scriptus-pricei/data/matlab/data_bushbuck_UM2.csv");
        %get the number of rows and cols
        n_rows = size(data_bushbuck,1);
        n_cols = size(data_bushbuck,2);
        
        %blank array storage for matrix for each image
        teeth_data_bushbuck = zeros(2,100,n_rows/2);
        
        
        %loop to assign each tooth's coordinates to the appropriate spot in the
        %array
        for i=1:2:n_rows
            %get index for the image we are on 
            j = find((1:2:n_rows)==i);
            
            %assign matrix for image j 
             X = data_bushbuck{i:(i+1), 2:n_cols};
            
            %resample so all of the curves have 100 points
            teeth_data_bushbuck(:,:,j) = ReSampleCurve(X,100);
        end



%Get the LM2's froms criptus and pricei
toothtype = {"UM2"};
species = {"pricei", "scriptus"};
for t=1
  data_scriptus = readtable(wd + "/teeth-scriptus-pricei/data/matlab/data_"+toothtype(t)+"_scriptus.csv");
  data_pricei = readtable(wd + "/teeth-scriptus-pricei/data/matlab/data_"+toothtype(t)+"_pricei.csv");
  data=[data_scriptus; data_pricei];
        
        %get the number of rows and cols
        n_rows = size(data,1);
        n_cols = size(data,2);
        
        %blank array storage for matrix for each image
        teeth_data = zeros(2,100,n_rows/2);
        
        
        %loop to assign each tooth's coordinates to the appropriate spot in the
        %array
        for i=1:2:n_rows
            %get index for the image we are on 
            j = find((1:2:n_rows)==i);
            
            %assign matrix for image j 
             X = data{i:(i+1), 2:n_cols};
            
            %resample so all of the curves have 100 points
            teeth_data(:,:,j) = ReSampleCurve(X,100);
        end        

    end



load(wd + "/teeth-scriptus-pricei/data/matlab/pairwise_distances_"+toothtype(t)+".mat","ddd");

size(ddd,1)
ddd_bushbuck_UM2 = ddd;

for i = 1:size(ddd,1);
    for j = 1:size(teeth_data_bushbuck,3);
        ddd_bushbuck_UM2(i,size(ddd,1)+j)  = GeodesicElasticClosed(teeth_data(:,:,i),teeth_data_bushbuck(:,:,j));     
        ddd_bushbuck_UM2(size(ddd,1)+j,i) = ddd_bushbuck_UM2(i,size(ddd,1)+j);
    end
end

for i = 1:size(teeth_data_bushbuck,3);
    for j = 1:size(teeth_data_bushbuck,3);
        ddd_bushbuck_UM2(size(ddd,1)+i,size(ddd,1)+j)  = GeodesicElasticClosed(teeth_data_bushbuck(:,:,i),teeth_data_bushbuck(:,:,j));     
        ddd_bushbuck_UM2(size(ddd,1)+j,size(ddd,1)+i) = ddd_bushbuck_UM2(size(ddd,1)+i,size(ddd,1)+j)
        
    end
end

save(wd + "/teeth-scriptus-pricei/data/matlab/pairwise_distances_bushbuck_UM2.mat","ddd_bushbuck_UM2")
  csvwrite(wd + "/teeth-scriptus-pricei/data/matlab/pairwise_distances_bushbuck_UM2.csv",ddd_bushbuck_UM2)